### DISCUSSION TITLE
2.7.1 C++ CPU PPOCR类内存泄漏

### DISCUSSION THREAD

**Discussion by Aquapisces** [PaddlePaddle/PaddleOCR#12398] [Q&A] [answered]
- 系统环境/System Environment：win10-x64, VisualStudio 2019
- 版本号/Version：Paddle：  PaddleOCR：2.7.1  问题相关组件/Related components：PPOCR类

请从这里获取我的工程：
https://github.com/Aquapisces/UsePaddleOcr271

VS2019打开可以直接运行，main.cpp中line13~21：
`	PPOCR* ocr = new PPOCR;
        ......
	// test memory leak
        delete ocr;
        ocr = nullptr;`
测试发现这里delete ocr后并未释放内存，我定位到是PaddleOCR\ocr_det.cpp的line61:
`  this->predictor_ = paddle_infer::CreatePredictor(config);`
执行这行后内存涨到650Mb左右，之后再也释放不掉。ocr_cls.cpp和ocr_rec.cpp同样也是paddle_infer::CreatePredictor后未能释放。
看你们代码里都是智能指针，按理说不会释放不了啊，百思不得其解

**Comment by jun20061588** [comment]
同样的问题

**Comment by gowy222** [comment]
指针这东西还真敢托付给智能...

**Comment by tink2123** [comment]
请问在不断的推理过程中，内存会持续增加吗？是否方便测试下单张图循环推理n次，内存的增长情况。

**Comment by Aquapisces** [comment]
> 请问在不断的推理过程中，内存会持续增加吗？是否方便测试下单张图循环推理n次，内存的增长情况。
光加载模型就已经有内存泄露了哇。
推理图片的话，推理个小尺寸图片，内存上去了下不来；再推理同样大小的图片，内存不继续涨了但也不下去；再推理个大尺寸的图片，内存再上去一些、然后也是下不来了；再推理个同样的大尺寸图片，内存也还是不继续涨了但也不下去。

**Comment by ShubhamZoop** [comment]
https://github.com/PaddlePaddle/PaddleOCR/issues/11639

**Comment by vivienfanghuagood** [comment, marked as answer]
你好，这个原因经过分析，本质上不属于内存泄漏，是Paddle框架将Tensor做了缓存复用，这部分内存 在下次遇到同样shape tensor的时候会复用，从而避免调用系统的allocator。如果对内存比较敏感，可以export FLAGS_allocator_strategy=naive_best_fit，这个会一定程度上缓解CPU的内存占用。后续我们将进一步优化Allocator的逻辑，以提供更合理的内存复用策略。

**Reply by chenxz11** [reply]
您好，想问一下。我这边编译paddleocr（gpu）的时候，在visual studio2022编译器环境下运行就是正常的，但是单独运行，就会出现内存泄露（持续测试一批图片）是什么原因呢？期待您的回复。
